internal/engine/experiment/quicping/quicping.go

```diff
@@ -0,0 +1,479 @@
+	ctx, cancel := context.WithTimeout(ctx, deadline)
+	defer cancel()
+
+	sendInfoChan := make(chan requestInfo)
+	recvInfoChan := make(chan responseInfo)
+	pingMap := make(map[string]*pingInfo)
+
+	// start sender and receiver goroutines
+	go m.sender(ctx, pconn, udpAddr, sendInfoChan, sess)
+	go m.receiver(ctx, pconn, recvInfoChan, sess)
+L:
+	for {
+		select {
+		case req := <-sendInfoChan: // a new ping was sent
+			if req.err != nil {
+				tk.Pings = append(tk.Pings, &SinglePing{
+					ConnIdDst:   req.dstID,
+					ConnIdSrc:   req.srcID,
+					Failure:     archival.NewFailure(req.err),
+					Request:     &model.ArchivalMaybeBinaryData{Value: string(req.raw)},
+					RequestTime: formatTime(req.t),
+				})
+				continue
+			}
+			pingMap[req.srcID] = &pingInfo{request: &req}
+
+		case resp := <-recvInfoChan: // a new response has been received
+			if resp.err != nil {
+				// resp failure means we cannot assign the response to a request
+				tk.UnexpectedResponses = append(tk.UnexpectedResponses, makeResponse(&resp))
+				continue
+			}
+			var (
+				ping *pingInfo
+				ok   bool
+			)
+			// match response to request
+			if ping, ok = pingMap[resp.dstID]; !ok {
+				// version negotiation response with an unknown destination ID
+				tk.UnexpectedResponses = append(tk.UnexpectedResponses, makeResponse(&resp))
+				continue
+			}
+			ping.responses = append(ping.responses, &resp)
+
+		case <-ctx.Done():
+			break L
+		}
+	}
+	// transform ping requests into TestKeys.Pings
+	timeoutErr := errors.New("i/o timeout")
+	for _, ping := range pingMap {
+		if ping.request == nil { // this should not happen
+			return errors.New("internal error: ping.request is nil")
+		}
+		if len(ping.responses) == 0 {
+			tk.Pings = append(tk.Pings, &SinglePing{
+				ConnIdDst:   ping.request.dstID,
+				ConnIdSrc:   ping.request.srcID,
+				Failure:     archival.NewFailure(timeoutErr),
+				Request:     &model.ArchivalMaybeBinaryData{Value: string(ping.request.raw)},
+				RequestTime: formatTime(ping.request.t),
+			})
+			continue
+		}
+		var responses []*SinglePingResponse
+		for _, resp := range ping.responses {
+			responses = append(responses, makeResponse(resp))
+		}
+		tk.Pings = append(tk.Pings, &SinglePing{
+			ConnIdDst:   ping.request.dstID,
+			ConnIdSrc:   ping.request.srcID,
+			Failure:     nil,
+			Request:     &model.ArchivalMaybeBinaryData{Value: string(ping.request.raw)},
+			RequestTime: formatTime(ping.request.t),
+			Responses:   responses,
+		})
+	}
+	return nil
+}
+
+// DissectVersionNegotiation dissects the Version Negotiation response.
+// It returns the supported versions and the destination connection ID of the response,
+// The destination connection ID of the response has to coincide with the source connection ID of the request.
+// https://www.rfc-editor.org/rfc/rfc9000.html#name-version-negotiation-packet
+func (m *Measurer) DissectVersionNegotiation(i []byte) ([]uint32, ConnectionID, error) {
```

Comment:
Please, move this function to a separate file. Logically, this is a basic building block for the main implementation of the experiment, so I'd like it to be in a separate file. The way I see it, our stack here is like:

```
  experiment (quicping.go)
  --------
  separate file (maybe quic.go?)
  --------
  crypto (crypto.go)

```

Reply:
maybe `dissect.go` ?